packages/client/src/index.test.ts

```diff
@@ -42,49 +31,68 @@ const discoverResponse = {
   end_session_endpoint: `${BASE_URL}/oidc/session/end`,
 };
 
-const fakeTokenResponse = {
-  access_token: 'access_token',
-  expires_in: 3600,
-  id_token: 'id_token',
-  refresh_token: 'refresh_token',
-};
-
 const generateIdToken = async () => {
   const { privateKey, publicKey } = await generateKeyPair('RS256');
 
   if (!(publicKey instanceof KeyObject)) {
     throw new TypeError('key is not instanceof KeyObject, check envirionment');
   }
 
-  const key = publicKey.export({ format: 'jwk' });
-  nock(BASE_URL)
-    .get('/oidc/jwks')
-    .reply(200, { keys: [key] });
-  const idToken = await new SignJWT({})
+  return new SignJWT({})
     .setProtectedHeader({ alg: 'RS256' })
     .setAudience(CLIENT_ID)
     .setSubject(SUBJECT)
-    .setIssuer(discoverResponse.issuer)
+    .setIssuer(ISSUER)
     .setIssuedAt()
     .setExpirationTime('2h')
     .sign(privateKey);
+};
 
-  return { idToken, key };
+const ID_TOKEN = generateIdToken();
+
+const fakeTokenResponse = {
+  access_token: 'accessToken1',
+  refresh_token: 'refreshToken1',
+  id_token: ID_TOKEN,
+  token_type: 'tokenType1',
+  scope: DEFAULT_SCOPE_STRING,
+  expires_in: 3600,
 };
 
-describe('LogtoClient', () => {
-  beforeEach(async () => {
-    const { idToken } = await generateIdToken();
+jest.mock('./api', () => {
+  const discover = jest.fn(async () => discoverResponse);
+  const grantTokenByAuthorizationCode = jest.fn(async () => Promise.resolve(fakeTokenResponse));
+  const grantTokenByRefreshToken = jest.fn(async () => Promise.resolve(fakeTokenResponse));
 
-    nock(BASE_URL)
-      .post('/oidc/token')
-      .reply(200, {
-        ...fakeTokenResponse,
-        id_token: idToken,
-      });
-    nock(BASE_URL).get('/oidc/.well-known/openid-configuration').reply(200, discoverResponse);
-  });
+  return {
+    discover,
+    grantTokenByAuthorizationCode,
+    grantTokenByRefreshToken,
+  };
+});
+
+const CODE_VERIFIER = 'codeVerifier1';
+const STATE = 'state1';
 
+jest.mock('./utils/assembler', () => {
```

Comment:
Need to mock because:

-   We just require it to return specific value as we expected here
 
    If we call the actual `getLoginUrlWithCodeVerifierAndState`,
    we have to mock `generateState` and `generateCodeVerifier` instead
    so that we force `getLoginUrlWithCodeVerifierAndState` to return same url, codeVerifier and state.
    
-   `getLoginUrlWithCodeVerifierAndState` should be completely tested in `assembler.test.ts` and then we just care what it does but not how to implement it.

    I'd better mock it here.

For reference
  
```ts
export const getLoginUrlWithCodeVerifierAndState = async (
  parameters: LoginPrepareParameters
): Promise<{ url: string; codeVerifier: string; state: string }> => {
  const { baseUrl, clientId, scope, redirectUri } = parameters;
  const codeVerifier = generateCodeVerifier();
  const codeChallenge = await generateCodeChallenge(codeVerifier);
  const state = generateState();

  const url = `${baseUrl}?${qs.stringify({
    client_id: clientId,
    scope,
    response_type: 'code',
    redirect_uri: [redirectUri],
    prompt: 'consent',
    code_challenge: codeChallenge,
    code_challenge_method: 'S256',
    state,
  })}`;

  return { url, codeVerifier, state };
};
```